File: bigtangle-typescript/src/net/bigtangle/core/BlockType.ts
import { NetworkParameters } from './NetworkParameters';

export enum BlockType {
    BLOCKTYPE_INITIAL,
    BLOCKTYPE_TRANSFER,
    BLOCKTYPE_REWARD,
    BLOCKTYPE_TOKEN_CREATION,
    BLOCKTYPE_USERDATA,
    BLOCKTYPE_CONTRACT_EVENT,
    BLOCKTYPE_GOVERNANCE,
    BLOCKTYPE_FILE,
    BLOCKTYPE_CONTRACT_EXECUTE,
    BLOCKTYPE_CROSSTANGLE,
    BLOCKTYPE_ORDER_OPEN,
    BLOCKTYPE_ORDER_CANCEL,
    BLOCKTYPE_ORDER_EXECUTE,
    BLOCKTYPE_CONTRACTEVENT_CANCEL
}

export interface BlockTypeConfig {
    allowCoinbaseTransaction: boolean;
    maxSize: number;
    requiresCalculation: boolean;
}

export const BlockTypeConfigs: { [key in BlockType]: BlockTypeConfig } = {
    [BlockType.BLOCKTYPE_INITIAL]: { allowCoinbaseTransaction: false, maxSize: Number.MAX_SAFE_INTEGER, requiresCalculation: false },
    [BlockType.BLOCKTYPE_TRANSFER]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_REWARD]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_REWARD_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_TOKEN_CREATION]: { allowCoinbaseTransaction: true, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_USERDATA]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_CONTRACT_EVENT]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_GOVERNANCE]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_FILE]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_CONTRACT_EXECUTE]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_CROSSTANGLE]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_ORDER_OPEN]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_ORDER_CANCEL]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_ORDER_EXECUTE]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
    [BlockType.BLOCKTYPE_CONTRACTEVENT_CANCEL]: { allowCoinbaseTransaction: false, maxSize: NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, requiresCalculation: false },
};

export function allowCoinbaseTransaction(type: BlockType): boolean {
    return BlockTypeConfigs[type].allowCoinbaseTransaction;
}

export function getMaxBlockSize(type: BlockType): number {
    return BlockTypeConfigs[type].maxSize;
}

export function requiresCalculation(type: BlockType): boolean {
    return BlockTypeConfigs[type].requiresCalculation;
}

File: bigtangle-core/src/main/java/net/bigtangle/core/BlockType.java
package net.bigtangle.core;

import net.bigtangle.params.NetworkParameters;

public enum BlockType {
	/**
	 * To add new BLOCKTYPES to implement type specific function The order can not
	 * be changed for history! enum cardinal is saved in database. It can be added
	 * new at the end of enum Type
	 */

	BLOCKTYPE_INITIAL(false, Integer.MAX_VALUE, false), // Genesis block
	BLOCKTYPE_TRANSFER(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // Default
	BLOCKTYPE_BEACON(false, NetworkParameters.MAX_REWARD_BLOCK_SIZE, false), // Rewards
	BLOCKTYPE_TOKEN_CREATION(true, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // tokenissuance
	BLOCKTYPE_USERDATA(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // User-defined-data
	BLOCKTYPE_CONTRACT_EVENT(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // Smart-contracts
	BLOCKTYPE_GOVERNANCE(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // Governance
	BLOCKTYPE_FILE(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // User-defined-file
		BLOCKTYPE_CROSSTANGLE(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // mainnet to permissioned
	BLOCKTYPE_ORDER_OPEN(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // new-order
	BLOCKTYPE_ORDER_CANCEL(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // cancel-order
	BLOCKTYPE_CONTRACTEVENT_CANCEL(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // Order execution
	BLOCKTYPE_STAKE(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // PoS staking deposit
	BLOCKTYPE_SLASHING(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false), // PoS slashing proof
	BLOCKTYPE_NFT(false, NetworkParameters.MAX_DEFAULT_BLOCK_SIZE, false); // NFT creation / user data

	private final boolean allowCoinbaseTransaction;
	private final int maxSize;
	private final boolean requiresCalculation;

	BlockType(boolean allowCoinbaseTransaction, int maxSize, boolean requiresCalculation) {
		this.allowCoinbaseTransaction = allowCoinbaseTransaction;
		this.maxSize = maxSize;
		this.requiresCalculation = requiresCalculation;
	}

	public boolean allowCoinbaseTransaction() {
		return allowCoinbaseTransaction;
	}

	public int getMaxBlockSize() {
		return maxSize;
	}

	public boolean requiresCalculation() {
		return requiresCalculation;
	}
}
